lib/spack/spack/database.py

```diff
@@ -1238,6 +1238,19 @@ def add(self, spec, directory_layout, explicit=False):
         with self.write_transaction():
             self._add(spec, directory_layout, explicit=explicit)
 
+    @_autospec
+    def _update_install_time(self, spec, absolute_time=None, delta=None):
```

Comment:
Extending the interface of database for something only used in tests is bad practice. I was hoping you could do something like

```
a = Spec("a dev_path=...")
b = Spec(b dev_path=...")
a.add_dependency_edge(b, ...)
a._mark_concrete()

db = Database(...)
db.add(b, install_time=2.0)
db.add(a, install_time=1.0)

assert get_ovewrite_specs([a, b], database=db) == ...
```

Reply:
The functionality added in this PR accesses at least two interfaces:

* The specs (to see if they are installed)
* The database

if you pass in a test instance DB, that must be used to mock the global DB, so that specs grab from that properly. In this case, I happen to know from being familiar with all the underlying logic that assigning this DB object you create to be the store would be sufficient (at least for now), but it is subject to conceptual drift (if the notion of how the spec is installed changes, or it reaches into any other part of the global state, then there are possibilities for more-insidious bugs).

So I like the general idea of producing a more-isolated test for this, but I don't think that can be done cleanly without mocking both the Spec object as well as the Database: does that seem reasonable? Another option along those lines is to promote the concept of install-time to the spec level (at which point the only object that needs to be mocked is the Spec), although that would conflict with your no-functions-in-objects-just-for-tests preference <sup>1</sup>.

Also regarding your comment about the test taking a long time, I think that could be improved in general without specifically changing this test. Originally I was going to blame the fixtures, but actually the longest running parts appear to be the calls to `install()`, I think this could be improved by mocking the filesystem. Overall that isn't necessarily something to address here (although it may still be worthwhile having this test in addition to the more-specific one), but the integration-level tests certainly have a place somewhere in our testing modules, and speeding them up would be worthwhile.

1. That being said, while I understand the rule of thumb in play there, I don't think it applies in this case: for example I see no difference between using a private function like `_mark_concrete` and using `_update_install_time`; it's true that `_mark_concrete` has an already-existing meaningful intent, but you can use it here because you understand it also has a meaningful public-API-like interpretation, and for that matter I can say the same thing about `_update_install_time`).

Reply:
The test now refactors environment to use an (optionally) injected Database (and `Database` now has an `installed` method).